### PR TITLE
[PBD interim fix] Convert <EligibilityWarnings> to TypeScript

### PR DESCRIPTION
**Pull request by p-jackson**
#38141 is going to require a number of changes to the `<EligibilityWarnings>` component. Converting to TS since I need to understand the data structures involved and might as well document them while I'm here.

#### Changes proposed in this Pull Request

* Convert `<EligibilityWarnings>` components to ts
* Add `isKnownHoldType` to handle case where API returned an unknown hold (previously this would break the render)
* Add some jsdoc annotations to selectors so the types flow through to the component
* Add `supportUrl: null` to a few places since TS wasn't happy checking for the existence of the property here https://github.com/Automattic/wp-calypso/blob/aaa6703f1d3e2663ecb69bf795b9f9dabdf9aeac/client/blocks/eligibility-warnings/hold-list.tsx#L141

Question: have we got a better way of dealing with `translate` not returning a `string` when we know it will in this case?
https://github.com/Automattic/wp-calypso/blob/aaa6703f1d3e2663ecb69bf795b9f9dabdf9aeac/client/blocks/eligibility-warnings/index.tsx#L73
Maybe `<Banner>`'s `title` prop should just accept a `ReactChild`.

#### Testing Instructions

To see the `<EligibiliyWarnings>` component, pick a site that doesn't support AT and try to upload a plugin or theme, or change the hosting options (biz plan only)

### CODE REVIEW

**Comment by adamziel** [comment]
> Question: have we got a better way of dealing with translate not returning a string when we know it will in this case?

One way would be something like:

```ts
const title = translate<string>( 'Business plan required' );
```

But that's not much of an improvement, is it? :-) Your idea of making `<Banner>`'s `title` prop accept a ReactChild is much better

**Comment by sirreal** [comment]
> Question: have we got a better way of dealing with translate not returning a string when we know it will in this case?

We don't know it will. This is due to translate being hookable, meaning arbitrary hooks can change the result.

When typing translate, I'd originally made the type more like `translate( s: string ): string; translate( s: string, { components: {} } ): ReactFragment`, but that turned out not to be accurate because of hooks.

I'd very much like to find a better solution 🙂

**Comment by p-jackson** [comment]
The `<Banner>` (and therefore the `title` variable) is being removed from `<EligibilityWarnings>` in a subsequent PR, so happy with just casting it to `string` here since it will soon become irrelevant :)

File: client/blocks/eligibility-warnings/hold-list.tsx [hunk 1]
```diff
@@ -112,34 +124,43 @@ export const HoldList = ( { holds, isPlaceholder, siteSlug, translate } ) => {
...
-								<div className="eligibility-warnings__action">
-									<Button
-										compact
-										href={ holdMessages[ hold ].supportUrl }
-										rel="noopener noreferrer"
-									>
-										{ translate( 'Help' ) }
-									</Button>
+					map( holds, hold =>
+						! isKnownHoldType( hold, holdMessages ) ? null : (
```
Comment (adamziel) [inline client/blocks/eligibility-warnings/hold-list.tsx]: Now it makes me think we should log something when `hold` isn't a known hold type :-)

**Review by adamziel — APPROVED** [review]
🚢

File: client/blocks/eligibility-warnings/hold-list.tsx (hunk 1 above)
Comment (p-jackson) [reply client/blocks/eligibility-warnings/hold-list.tsx]: Yeah with TS making us acknowledge where all the runtime errors are, it makes me think we should use something like sentry.io

**Pull request merged**